### ISSUE TITLE
Webworker in IE11 and Edge does not work and stalls. Possible issue with support for canvas data. 

### ISSUE THREAD

**Issue by danrossi** [nodeca/pica#57]
Both browsers in Windows 10 fails to get past 

```
Resize buffer in WebWorker
```

Unfortunately if I disable web workers like in IE11 on Windows 7 the entire browser freezes until it's finished running the massive loop for the resizing. 

```
window.pica.resizeCanvas(this.srcCanvas, this.dstCanvas, this.picaConfig , this.onBlobResize.bind(this));
```

My code looks like this. the source canvas is not an image but a html5mediaelement. 

If I enable WebGL Edge works fine but if it's disabled and tries to use webworkers it stalls. No callback happens and no log is sent. 

If WebGL is enabled IE11 on Windows 10 tries to use that but I get a black canvas so that is broken on IE11 on Windows 10. 

Is WebGL stable to use now in terms of the aliasing ? I  don't seem to see this aliasing problem at all. 

I guess I have to be selective what can use WebGL but I also need to figure out why any of the IE browsers do not like Webworkers either.

**Comment by puzrin** [comment]
> Is WebGL stable to use now in terms of the aliasing ?

No. I strongly NOT recommend to use it in production, it's really buggy.

Make sure you use latest build (or browserify with fresh dependencies), see https://github.com/nodeca/pica/issues/55 for IE bugs. It passes typed arrays to worker, not fresh features like canvases. That should work. 

Since i'm not IE user, any help about IE+webworkers will be appreciated.

**Comment by puzrin** [comment]
PS. I'd say you tryed to go in wrong direction. Right direction is to make webworkers ok in IE/Win or fail correctly. Need more info why problem happens.

**Comment by danrossi** [comment]
I'll try and debug it but the issue is no logging or error returned, no callback etc. 

What do you mean wrong way ? I was testing WebGL on Edge and it's perfectly fine. With it disabled webworkers is stalled. Same on IE11. 

I've tested WebGL on chrome in OSX and I can't see anything wrong with aliasing either. 

So do I need to try and install dependancies and rebuild ?

**Comment by puzrin** [comment]
webgl = wrong way, because it's buggy and not supported.

If you use browserified version - make sure you use latest from `dist/` folder. If you use npm modules - make sure latest pica installed. After that's done - check that webworkers still have problems in win10/IE

**Comment by danrossi** [comment]
I've just been testing out WebGL. I see no issues with it at all. on microsoft edge browser and IE11 on windows 10 and Chrome on OSX. I'll do more testing with that. 

In the demo , webworkers is working for some reason as in I get a fallback. But in mine I get no callback. The source canvas is not added to the DOM however i'm drawing to it from a video element. I'll try my best to debug it. 

I have rebuilt the files from github sources. There is lacking documentation there in terms on how to run the build. I was able to to that after fixing the npm install as one of the deps has problems on osx . 

```
make browserify
```

**Comment by puzrin** [comment]
@danrossi there are hangs on some mobiles. I can't prohibit you to use webgl :). But you will be without help if something happen with it.

---

1. https://github.com/nodeca/pica/tree/master/dist - those files are prebuilt, you can use those directly. See https://github.com/nodeca/pica/blob/master/CHANGELOG.md, the only change last 2 versions were related to `webworkify` package update
2. Previous demo was with pica 2.0.2. Just updated it to use pica 2.0.4, please check.

**Comment by danrossi** [comment]
It seems the rebuild on my end possibly related to webworkify is working correctly now sorry about that. The builds from the sources were not. 

So IE11 on both Windows 7 and 10 and Edge browser are ok now. It defaults to work worker ok. I'll close this but if there is a ticket regards to the WebGL I can show some examples of it working.

**Comment by danrossi** [comment]
WebGL on mobile is sketchy. ie Chrome on Android 4.3 on an older Samsung S3. Firefox is ok. no idea about stock browser. If that is what you mean It can be selectively enabled. 

No idea about IOS but I will check. About to get to that part now. 

I've had nothing but trouble with Chrome on an older Samsung in regards to rendering webgl from video textures. So I believe is the graphics in the device. Later models are ok. There is a flag in there called something like "media pipeline" which makes it work but it causes severe frame dropping on any video.

**Comment by puzrin** [comment]
https://github.com/d08ble/pica/tree/webgl-dynamic-shader there are completely rewritten shaders. If you have possibility to organize testing on wide range of devices, i can kick @d08ble to rebase code and create a better build for you.

That code should fix issues with aliasing and missed lines.

**Comment by puzrin** [comment]
I remember problem, that webgl returned broken image on `xiaomi redme note`, and that sometime it was impossible to detect bad result and do fallback to safe methods.

**Comment by danrossi** [comment]
Gotcha. I'll check out that branch and do some testing. It might be best to just not enable webgl for Android. webworkers seem to do a fine job on desktop anyway and just tested on Ipad and its fine. Yet to test Android out. The webworkers seemed like there was still a delay under resource destrictions like in vmware though  ? Maybe 5-10 seconds though.  

I don't have access to many devices I'm afraid but an S3 , S7 and the emulators. That doesn't help test graphics chips and webgl sadly. I know for sure Chromium people only test on Nexus devices ! So I bet they all work fine.

**Comment by danrossi** [comment]
Sorry to update you here. It seemed like it was working. It even worked on both Android Xamarin on Chrome and Firefox. IE11 Windows 7 and Edge browser and IE11 on Windows 10 are fine. 

It is crashing for Chrome in both Windows 7 and Windows 10. I checked again and Chrome was ok on Windows 7. There is an obvious delay firing up webgl on first page load. 

It is crashing on Firefox on Windows 10 with this error

```
Error: WebGL: Disallowing antialiased backbuffers due to blacklisting
```

The windows machines are vmwares though.  But then so is the Android emulator. 

It is working on IOS / Ipad 3. The latency is almost 2 seconds with webgl and 5 seconds for web workers. 

I just checked the Samsung S3 Android 4.3. Chrome is fine with repeated requests but latency is parity with web workers because its graphics is bad. Firefox crashed after 3 resize attempts. Latency was 5 seconds. web workers 3 seconds.
